tests/interlocked_win32_ut/interlocked_win32_ut.c

```diff
@@ -0,0 +1,708 @@
+    int64_t return_val = interlocked_exchange_64(&target, value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int64_t, INT64_MIN, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_033: [interlocked_exchange_8 shall call InterlockedExchange8 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_034 : [interlocked_exchange_8 shall return the initial value pointed to by target.]*/
+TEST_FUNCTION(interlocked_exchange_8_calls_InterlockedExchange8)
+{
+    ///arrange
+    volatile int8_t target = INT8_MIN;
+    int8_t value = INT8_MAX;
+    STRICT_EXPECTED_CALL(mock_InterlockedExchange8((volatile char*)&target, (char)value))
+        .SetReturn(INT8_MIN);
+
+    ///act
+    int8_t return_val = interlocked_exchange_8(&target, value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int8_t, INT8_MIN, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_035: [interlocked_exchange_add shall call InterlockedExchangeAdd from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_036 : [interlocked_exchange_add shall return the initial value of * addend.]*/
+TEST_FUNCTION(interlocked_exchange_add_calls_InterlockedExchangeAdd)
+{
+    ///arrange
+    volatile int32_t addend = INT32_MIN;
+    int32_t value = INT32_MAX;
+    STRICT_EXPECTED_CALL(mock_InterlockedExchangeAdd((volatile LONG*)&addend, (LONG)value))
+        .SetReturn(INT32_MIN);
+
+    ///act
+    int32_t return_val = interlocked_exchange_add(&addend, value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int32_t, INT32_MIN, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_037: [interlocked_exchange_add_64 shall call InterlockedExchangeAdd64 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_038 : [interlocked_exchange_add_64 shall return the initial value of * addend.]*/
+TEST_FUNCTION(interlocked_exchange_add_64_calls_InterlockedExchangeAdd64)
+{
+    ///arrange
+    volatile int64_t addend = INT64_MIN;
+    int64_t value = INT64_MAX;
+    STRICT_EXPECTED_CALL(mock_InterlockedExchangeAdd64((volatile LONG64*)&addend, (LONG64)value))
+        .SetReturn(INT64_MIN);
+
+    ///act
+    int64_t return_val = interlocked_exchange_add_64(&addend, value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int64_t, INT64_MIN, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_039: [interlocked_exchange_pointer shall call InterlockedExchangePointer from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_040 : [interlocked_exchange_pointer shall return the initial address pointed to by the target parameter]*/
+TEST_FUNCTION(interlocked_exchange_pointer_calls_InterlockedExchagePointer)
+{
+    ///arrange
+    int value1 = 1;
+    int value2 = 2;
+    void* volatile target = &value1;
```

Comment:
>void* volatile target [](start = 4, length = 21)

Interesting, this pointer looks as I expect it to look. 

As opposed to a pointer from integration tests, which looked the opposite.

One of them **must** be wrong!

Reply:
Yeah the 'volatile void**' is wrong.

---
In reply to: [427745365](https://github.com/Azure/azure-c-util/pull/43#discussion_r427745365) [](ancestors = 427745365)